### ISSUE TITLE
Numpy dtype disjunction doesn't work like type disjunction?

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#4260]
```python
from typing import Any
import numpy as np
import numpy.typing as npt

NumpyIntegralArray = npt.NDArray[np.integer[Any]]
NumpyRealArray = npt.NDArray[np.floating[Any]]

# This works!
# def smooth_data(values: NumpyRealArray) -> None:

# This doesn't!
def smooth_data(values: NumpyIntegralArray | NumpyRealArray) -> None:
    pass

total_observed = np.zeros(10)
smooth_data(np.asarray(total_observed))
```

**Comment by erictraut** [comment]
This occurs because of bidirectional type inference. Here's a simplified sample:

```python
from typing import Generic, TypeVar
from numpy import dtype, generic, integer, floating

_DT = TypeVar("_DT", covariant=True, bound=dtype)
_SCT = TypeVar("_SCT", bound=generic)

class nd(Generic[_DT]): ...

def asarray(a: nd[dtype[_SCT]]) -> nd[dtype[_SCT]]: ...

def func1(values: nd[dtype[integer]] | nd[dtype[floating]]) -> None:
    pass

def func2(v: nd[dtype[floating]]):
    func1(asarray(v))
```

The evaluated return type of `asarray` is affected by the expected type of the `values` parameter for `func1`. In this case, the expected type is `nd[dtype[integer]] | nd[dtype[floating]]` which means that the expected return type of `asarray(v)` is `nd[dtype[integer | floating]]`, which means that the expected type of the argument corresponding to argument `a` in `asarray` is `nd[dtype[integer | floating]]`, but it is receiving an argument of type `nd[dtype[floating]]`, which is not type compatible.

I'll need to think more about whether there's a way to prevent this error without breaking many other common cases.

The workaround is to assign the return value of `np.asarray(total_observed)` to a separate variable before passing it to `smooth_data` because this eliminates the use of bidirectional type inference.

**Comment by matangover** [comment]
Why is bidirectional inference kicking in if the type of `v` in not ambiguous?

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.286, which I just published. It will also be included in a future version of pylance.
